task21/js/main.js

```diff
@@ -0,0 +1,154 @@
+import '../css/style.css'
+import { createElementWithClassName } from './utils/createElement'
+import { renderLoading } from './modules/loading'
+import { removeLoading } from './modules/loading'
+import { renderErrorMessage } from './modules/error'
+
+const body = document.getElementById("js-body");
+const parent = document.getElementById("js-parent");
+
+const getFetchData = async (endpoint) => {
+    const response = await fetch(endpoint);
+
+    if (!response.ok){
+        console.error(`${response.status}:${response.statusText}`);
+        renderErrorMessage(parent, "問題が発生し表示することができません。");
+    }
+    return await response.json();
+};
+
+const fetchTableData = () => {
+    return new Promise((resolve) => {
+        setTimeout(() => {
+            resolve(getFetchData("https://mocki.io/v1/159d72ba-bd6b-429f-a28c-713eeaba9f72"));
+        }, 3000);
+        // setTimeout(() => {
+        //     resolve(getFetchData("https://mocki.io/v1/f12ae2d1-310d-4120-8749-47773d65e236"));
+        // }, 3000);//空配列
+        // setTimeout(() => {
+        //     resolve(getFetchData("https://httpstat.us/503"));
+        // }, 3000);//503error
+    });
+};
+
+const getUserData = async () => {
+    renderLoading(body);
+    try {
+        const json = await fetchTableData();
+        const data= json.data;
+        if(data.length === 0){
+            renderErrorMessage(parent, "データがありません。");
+            return;
+        }
+        return data;
+    } catch(error) {
+        renderErrorMessage(parent, "問題が発生し表示することができません。");
+    } finally {
+        removeLoading();
+    }    
+};
+
+const initialize = async () => {
+    const userContentsData = await getUserData();
+    userContentsData && renderTableContents(userTableColumn, userContentsData);
+    addEventForSortButton(userContentsData);
+}
```

Comment:
IMO
I think `addEventForSortButton(userContentsData);` should be in an if statement when you execute it here. Because if there is no userContentsData and the Table is not rendered, it will cause an error.

Or maybe it would be better to do it in `renderTableContents()`  instead of here.
```suggestion
const initialize = async () => {
    const userContentsData = await getUserData();
    if(userContentsData) {
      renderTableContents(userTableColumn, userContentsData);
      addEventForSortButton(userContentsData);
    }
}
```

Reply:
Thank you for your very informative comments.
I made this correction because I felt it was more appropriate to do it with renderTableContents than to do it here.
[bef43a8](https://github.com/seeeen044/JavaScript_handson/pull/22/commits/bef43a8434507975a303fe91a5639d2a3716e0c2)